messages/helpers.go

```diff
@@ -0,0 +1,187 @@
+		return nil
+	}
+
+	preprepareData, _ := proposalMessage.Payload.(*proto.Message_PreprepareData)
+
+	return preprepareData.PreprepareData.Proposal
+}
+
+// ExtractProposalHash extracts the proposal hash from the passed in message
+func ExtractProposalHash(proposalMessage *proto.Message) []byte {
+	if proposalMessage.Type != proto.MessageType_PREPREPARE {
+		return nil
+	}
+
+	preprepareData, _ := proposalMessage.Payload.(*proto.Message_PreprepareData)
+
+	return preprepareData.PreprepareData.ProposalHash
+}
+
+// ExtractRoundChangeCertificate extracts the RCC from the passed in message
+func ExtractRoundChangeCertificate(proposalMessage *proto.Message) *proto.RoundChangeCertificate {
+	if proposalMessage.Type != proto.MessageType_PREPREPARE {
+		return nil
+	}
+
+	preprepareData, _ := proposalMessage.Payload.(*proto.Message_PreprepareData)
+
+	return preprepareData.PreprepareData.Certificate
+}
+
+// ExtractPrepareHash extracts the prepare proposal hash from the passed in message
+func ExtractPrepareHash(prepareMessage *proto.Message) []byte {
+	if prepareMessage.Type != proto.MessageType_PREPARE {
+		return nil
+	}
+
+	prepareData, _ := prepareMessage.Payload.(*proto.Message_PrepareData)
+
+	return prepareData.PrepareData.ProposalHash
+}
+
+// ExtractLatestPC extracts the latest PC from the passed in message
+func ExtractLatestPC(roundChangeMessage *proto.Message) *proto.PreparedCertificate {
+	if roundChangeMessage.Type != proto.MessageType_ROUND_CHANGE {
+		return nil
+	}
+
+	rcData, _ := roundChangeMessage.Payload.(*proto.Message_RoundChangeData)
+
+	return rcData.RoundChangeData.LatestPreparedCertificate
+}
+
+// ExtractLastPreparedProposedBlock extracts the latest prepared proposed block from the passed in message
+func ExtractLastPreparedProposedBlock(roundChangeMessage *proto.Message) []byte {
+	if roundChangeMessage.Type != proto.MessageType_ROUND_CHANGE {
+		return nil
+	}
+
+	rcData, _ := roundChangeMessage.Payload.(*proto.Message_RoundChangeData)
+
+	return rcData.RoundChangeData.LastPreparedProposedBlock
+}
+
+// HasUniqueSenders checks if the messages have unique senders
+func HasUniqueSenders(messages []*proto.Message) bool {
+	if len(messages) < 1 {
+		return false
+	}
+
+	senderMap := make(map[string]struct{})
+
+	for _, message := range messages {
+		key := string(message.From)
+		if _, exists := senderMap[key]; exists {
+			return false
+		}
+
+		senderMap[key] = struct{}{}
+	}
+
+	return true
+}
+
+// HaveSameProposalHash checks if the messages have the same proposal hash
+func HaveSameProposalHash(messages []*proto.Message) bool {
+	if len(messages) < 1 {
+		return false
+	}
+
+	var hash []byte = nil
+
+	for _, message := range messages {
+		var extractedHash []byte
+
+		switch message.Type {
+		case proto.MessageType_PREPREPARE:
+			extractedHash = ExtractProposalHash(message)
+		case proto.MessageType_PREPARE:
+			extractedHash = ExtractPrepareHash(message)
+		default:
+			return false
+		}
+
+		if hash == nil {
```

Comment:
Please add comment to describe why you can set hash on the first loop

Reply:
Added:

[fb6dc59](https://github.com/0xPolygon/go-ibft/pull/21/commits/fb6dc59fe6864e70e8232c0f87bf49e540fbc208)